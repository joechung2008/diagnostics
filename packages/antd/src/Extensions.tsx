import { Menu, type MenuProps } from "antd";
import { byKey, isExtensionInfo, toNavLink } from "./utils";

const Extensions: React.FC<ExtensionsProps> = ({ extensions, onLinkClick }) => {
  const links = Object.values(extensions)
    .filter(isExtensionInfo)
    .map(toNavLink)
    .sort(byKey);

  const items = links.map((link) => ({
    key: link.key,
    label: link.name,
  }));

  const handleMenuClick: MenuProps["onClick"] = (info) => {
    const link = links.find(({ key }) => key === info.key);
    if (link) {
      onLinkClick?.(info.domEvent, link);
    }
  };

  return (
    <div className="extension-root">
      <Menu
        aria-label="Extensions"
        items={items}
        mode="vertical"
        onClick={handleMenuClick}
      />
    </div>
  );
};

export default Extensions;
