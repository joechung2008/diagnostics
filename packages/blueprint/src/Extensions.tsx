import { Tab, Tabs } from "@blueprintjs/core";
import { useState } from "react";
import { byKey, isExtensionInfo, toNavLink } from "./utils";

const Extensions: React.FC<ExtensionsProps> = ({ extensions, onLinkClick }) => {
  const links = Object.values(extensions)
    .filter(isExtensionInfo)
    .map(toNavLink)
    .sort(byKey);

  const [selectedTabId, setSelectedTabId] = useState<string>();

  return (
    <div className="extension-root">
      <Tabs
        vertical
        selectedTabId={selectedTabId}
        onChange={(tabId: string) => {
          setSelectedTabId(tabId);
          const link = links.find((l) => l.key === tabId);
          if (link) {
            onLinkClick(link);
          }
        }}
      >
        {links.map((link) => (
          <Tab
            key={link.key}
            aria-controls="extension"
            id={link.key}
            title={link.name}
          />
        ))}
      </Tabs>
    </div>
  );
};

export default Extensions;
