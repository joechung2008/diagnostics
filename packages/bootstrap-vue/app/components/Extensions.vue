<template>
  <nav class="extensions" aria-label="Extensions">
    <div class="d-flex flex-column gap-3">
      <BLink
        v-for="link in links"
        :key="link.key"
        href="#"
        class="text-start"
        @click="handleClick(link)"
      >
        {{ link.name }}
      </BLink>
    </div>
  </nav>
</template>

<script setup lang="ts">
import type { ExtensionsProps, KeyedNavLink } from "@/types/ExtensionsProps";
import { byKey, isExtensionInfo, toNavLink } from "@/utils/extensions";

const props = defineProps<ExtensionsProps>();

const links = computed(() =>
  Object.values(props.extensions ?? {})
    .filter(isExtensionInfo)
    .map(toNavLink)
    .sort(byKey)
);

function handleClick(link: KeyedNavLink) {
  props.onLinkClick?.(link);
}
</script>

<style scoped>
.extensions {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
  max-height: max-content;
  overflow-x: hidden;
  overflow-y: auto;
  padding: 0.5rem;
}
</style>
