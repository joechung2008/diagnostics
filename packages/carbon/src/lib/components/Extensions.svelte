<script lang="ts">
	import { byKey, isExtensionInfo, toNavLink } from '$lib';
	import { SideNavLink } from 'carbon-components-svelte';

	let { extensions, onLinkClick }: ExtensionsProps = $props();

	let links = $derived(
		Object.values(extensions).filter(isExtensionInfo).map(toNavLink).sort(byKey)
	);
</script>

{#each links as link (link.key)}
	<SideNavLink text={link.name} on:click={() => onLinkClick(link)} />
{/each}
