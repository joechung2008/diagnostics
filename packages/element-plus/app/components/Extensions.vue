<template>
    <nav class="extensions" aria-label="Extensions">
        <el-button v-for="link in links" :key="link.key" class="extension" @click="handleLinkClick(link)">
            {{ link.name }}
        </el-button>
    </nav>
</template>

<script setup lang="ts">
import type { ExtensionsProps, KeyedNavLink } from "@/types"
import { byKey, isExtensionInfo, toNavLink } from "../utils/extensions"

const props = defineProps<ExtensionsProps>()

const links = computed(() =>
    Object.values(props.extensions)
        .filter(isExtensionInfo)
        .map(toNavLink)
        .sort(byKey)
)

function handleLinkClick(link: KeyedNavLink) {
    props.onLinkClick(link)
}
</script>

<style scoped>
.extensions {
    box-sizing: border-box;
    display: flex;
    flex-direction: column;
    gap: 0.5rem;
    height: 100%;
    overflow-y: auto;
    padding-right: 0.5rem;
}

/* Override Element Plus default margin */
.extension {
    justify-content: start;
    margin: 0;
}
</style>
