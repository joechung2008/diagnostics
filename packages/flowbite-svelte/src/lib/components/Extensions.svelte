<script lang="ts">
  import { byKey, isExtensionInfo, toNavLink } from "$lib/utils";
  import { Sidebar, SidebarGroup, SidebarItem } from "flowbite-svelte";

  let { extensions, onLinkClick }: ExtensionsProps = $props();

  const links = $derived.by(() =>
    Object.values(extensions).filter(isExtensionInfo).map(toNavLink).sort(byKey)
  );
</script>

<div class="flex flex-none flex-col gap-0.5 overflow-y-auto p-1">
  <Sidebar alwaysOpen aria-label="Extensions" position="static" class="!z-0 w-auto">
    <SidebarGroup>
      {#each links as link (link.key)}
        <SidebarItem
          label={link.name}
          onclick={() => onLinkClick?.(link)}
          class="cursor-pointer"
          tabindex={0}
        />
      {/each}
    </SidebarGroup>
  </Sidebar>
</div>
