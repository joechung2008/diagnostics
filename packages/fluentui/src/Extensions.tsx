import { Button, makeStyles, mergeClasses } from "@fluentui/react-components";
import { useMemo } from "react";
import { byKey, isExtensionInfo, toNavLink } from "./utils";

const useStyles = makeStyles({
  navButton: {
    justifyContent: "flex-start",
    textAlign: "left",
    width: "100%",
    minHeight: "max-content",
    boxSizing: "border-box",
  },
});
const Extensions: React.FC<ExtensionsProps> = ({ extensions, onLinkClick }) => {
  const styles = useStyles();
  const links = useMemo(
    () =>
      Object.values(extensions)
        .filter(isExtensionInfo)
        .map(toNavLink)
        .sort(byKey),
    [extensions]
  );

  return (
    <nav className="extension-root" aria-label="Extensions">
      {links.map((link) => (
        <Button
          key={link.key}
          className={mergeClasses("extension-nav-button", styles.navButton)}
          appearance="subtle"
          onClick={(e) => onLinkClick?.(e, link)}
        >
          {link.name}
        </Button>
      ))}
    </nav>
  );
};

export default Extensions;
