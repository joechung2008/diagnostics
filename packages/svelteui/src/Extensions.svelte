<script lang="ts">
  import { Button } from '@svelteuidev/core'
  import { createEventDispatcher } from 'svelte'
  import type {
    Extension,
    ExtensionInfo,
    ExtensionsProps,
    KeyedNavLink
  } from './types'

  export let extensions: ExtensionsProps['extensions']
  export let onLinkClick: ExtensionsProps['onLinkClick'] | undefined

  const dispatch = createEventDispatcher<{
    linkClick: { event: Event; link: KeyedNavLink }
  }>()

  // Utility functions
  function isExtensionInfo(extension: Extension): extension is ExtensionInfo {
    return 'extensionName' in extension
  }

  function toNavLink(extension: ExtensionInfo): KeyedNavLink {
    return {
      key: extension.extensionName,
      name: extension.extensionName,
      url: extension.extensionName
    }
  }

  function byKey(a: KeyedNavLink, b: KeyedNavLink): number {
    return a.key.localeCompare(b.key)
  }

  $: links = Object.values(extensions)
    .filter(isExtensionInfo)
    .map(toNavLink)
    .sort(byKey)

  function handleClick(event: CustomEvent<unknown>, link: KeyedNavLink) {
    onLinkClick?.(event, link)
    dispatch('linkClick', { event, link })
  }
</script>

<nav aria-label="Extensions">
  {#each links as link (link.key)}
    <Button
      class="extension-link"
      variant="subtle"
      on:click={(event) => handleClick(event, link)}
    >
      {link.name}
    </Button>
  {/each}
</nav>

<style scoped>
  nav {
    display: flex;
    flex-direction: column;
    gap: 0.5rem;
    overflow: hidden auto;
  }
</style>
