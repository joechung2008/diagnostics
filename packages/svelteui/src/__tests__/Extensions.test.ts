import { fireEvent, render } from '@testing-library/svelte'
import { describe, expect, it, vi } from 'vitest'
import Extensions from '../Extensions.svelte'
import type { Extension, ExtensionError, ExtensionInfo } from '../types'

describe('Extensions', () => {
  it('should render single extension info', () => {
    const extensions: Record<string, Extension> = {
      'test-ext': {
        extensionName: 'TestExtension'
      } as ExtensionInfo
    }

    const { container } = render(Extensions, {
      props: {
        extensions,
        onLinkClick: undefined
      }
    })

    expect(container).toMatchSnapshot()
  })

  it('should render multiple extension infos sorted alphabetically', () => {
    const extensions: Record<string, Extension> = {
      'zebra-ext': {
        extensionName: 'ZebraExtension'
      } as ExtensionInfo,
      'alpha-ext': {
        extensionName: 'AlphaExtension'
      } as ExtensionInfo,
      'beta-ext': {
        extensionName: 'BetaExtension'
      } as ExtensionInfo
    }

    const { container } = render(Extensions, {
      props: {
        extensions,
        onLinkClick: undefined
      }
    })

    expect(container).toMatchSnapshot()
  })

  it('should filter out extension errors and only show extension infos', () => {
    const extensions: Record<string, Extension> = {
      'good-ext': {
        extensionName: 'GoodExtension'
      } as ExtensionInfo,
      'bad-ext': {
        lastError: {
          errorMessage: 'Failed to load',
          time: '2025-09-13T12:00:00Z'
        }
      } as ExtensionError,
      'another-good-ext': {
        extensionName: 'AnotherGoodExtension'
      } as ExtensionInfo
    }

    const { container } = render(Extensions, {
      props: {
        extensions,
        onLinkClick: undefined
      }
    })

    expect(container).toMatchSnapshot()
  })

  it('should render extensions with full extension info including config and stage definition', () => {
    const extensions: Record<string, Extension> = {
      'full-ext': {
        extensionName: 'FullExtension',
        config: {
          setting1: 'value1',
          setting2: 'value2'
        },
        stageDefinition: {
          dev: ['localhost'],
          prod: ['production.com']
        }
      } as ExtensionInfo
    }

    const { container } = render(Extensions, {
      props: {
        extensions,
        onLinkClick: undefined
      }
    })

    expect(container).toMatchSnapshot()
  })

  it('should render empty extensions object', () => {
    const extensions: Record<string, Extension> = {}

    const { container } = render(Extensions, {
      props: {
        extensions,
        onLinkClick: undefined
      }
    })

    expect(container).toMatchSnapshot()
  })

  it('should render only extension errors (no visible links)', () => {
    const extensions: Record<string, Extension> = {
      error1: {
        lastError: {
          errorMessage: 'Extension failed',
          time: '2025-09-13T12:00:00Z'
        }
      } as ExtensionError,
      error2: {
        lastError: {
          errorMessage: 'Another failure',
          time: '2025-09-13T13:00:00Z'
        }
      } as ExtensionError
    }

    const { container } = render(Extensions, {
      props: {
        extensions,
        onLinkClick: undefined
      }
    })

    expect(container).toMatchSnapshot()
  })

  it('should handle extensions with special characters in names', () => {
    const extensions: Record<string, Extension> = {
      'special-ext': {
        extensionName: 'Special_Extension-v2.1.0'
      } as ExtensionInfo,
      'another-special': {
        extensionName: 'Another@Special#Extension'
      } as ExtensionInfo
    }

    const { container } = render(Extensions, {
      props: {
        extensions,
        onLinkClick: undefined
      }
    })

    expect(container).toMatchSnapshot()
  })

  it('should handle onLinkClick prop', () => {
    const mockOnLinkClick = vi.fn()
    const extensions: Record<string, Extension> = {
      'clickable-ext': {
        extensionName: 'ClickableExtension'
      } as ExtensionInfo
    }

    const { container, getByText } = render(Extensions, {
      props: {
        extensions,
        onLinkClick: mockOnLinkClick
      }
    })

    expect(container).toMatchSnapshot()

    // Test click functionality
    const button = getByText('ClickableExtension')
    fireEvent.click(button)

    expect(mockOnLinkClick).toHaveBeenCalledTimes(1)
    expect(mockOnLinkClick).toHaveBeenCalledWith(
      expect.any(Event),
      expect.objectContaining({
        key: 'ClickableExtension',
        name: 'ClickableExtension',
        url: 'ClickableExtension'
      })
    )
  })
})
