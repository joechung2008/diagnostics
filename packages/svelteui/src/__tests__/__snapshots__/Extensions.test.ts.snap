// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Extensions > should filter out extension errors and only show extension infos 1`] = `
<div>
  <nav
    aria-label="Extensions"
    class="svelte-4iy04k"
  >
    
    
     
    <button
      class="extension-link svelteui-c-jlBcFc svelteui-Button-root svelteui-c-jlBcFc-cWGvDy-variation-subtle svelteui-c-jlBcFc-iPJLV-css svelte-5xpm5q"
      tabindex="0"
    >
       
      AnotherGoodExtension
       
       
       
    </button>
    
    
    
     
    <button
      class="extension-link svelteui-c-jlBcFc svelteui-Button-root svelteui-c-jlBcFc-cWGvDy-variation-subtle svelteui-c-jlBcFc-iPJLV-css svelte-5xpm5q"
      tabindex="0"
    >
       
      GoodExtension
       
       
       
    </button>
    
  </nav>
</div>
`;

exports[`Extensions > should handle extensions with special characters in names 1`] = `
<div>
  <nav
    aria-label="Extensions"
    class="svelte-4iy04k"
  >
    
    
     
    <button
      class="extension-link svelteui-c-jlBcFc svelteui-Button-root svelteui-c-jlBcFc-cWGvDy-variation-subtle svelteui-c-jlBcFc-iPJLV-css svelte-5xpm5q"
      tabindex="0"
    >
       
      Another@Special#Extension
       
       
       
    </button>
    
    
    
     
    <button
      class="extension-link svelteui-c-jlBcFc svelteui-Button-root svelteui-c-jlBcFc-cWGvDy-variation-subtle svelteui-c-jlBcFc-iPJLV-css svelte-5xpm5q"
      tabindex="0"
    >
       
      Special_Extension-v2.1.0
       
       
       
    </button>
    
  </nav>
</div>
`;

exports[`Extensions > should handle onLinkClick prop 1`] = `
<div>
  <nav
    aria-label="Extensions"
    class="svelte-4iy04k"
  >
    
    
     
    <button
      class="extension-link svelteui-c-jlBcFc svelteui-Button-root svelteui-c-jlBcFc-cWGvDy-variation-subtle svelteui-c-jlBcFc-iPJLV-css svelte-5xpm5q"
      tabindex="0"
    >
       
      ClickableExtension
       
       
       
    </button>
    
  </nav>
</div>
`;

exports[`Extensions > should render empty extensions object 1`] = `
<div>
  <nav
    aria-label="Extensions"
    class="svelte-4iy04k"
  />
</div>
`;

exports[`Extensions > should render extensions with full extension info including config and stage definition 1`] = `
<div>
  <nav
    aria-label="Extensions"
    class="svelte-4iy04k"
  >
    
    
     
    <button
      class="extension-link svelteui-c-jlBcFc svelteui-Button-root svelteui-c-jlBcFc-cWGvDy-variation-subtle svelteui-c-jlBcFc-iPJLV-css svelte-5xpm5q"
      tabindex="0"
    >
       
      FullExtension
       
       
       
    </button>
    
  </nav>
</div>
`;

exports[`Extensions > should render multiple extension infos sorted alphabetically 1`] = `
<div>
  <nav
    aria-label="Extensions"
    class="svelte-4iy04k"
  >
    
    
     
    <button
      class="extension-link svelteui-c-jlBcFc svelteui-Button-root svelteui-c-jlBcFc-cWGvDy-variation-subtle svelteui-c-jlBcFc-iPJLV-css svelte-5xpm5q"
      tabindex="0"
    >
       
      AlphaExtension
       
       
       
    </button>
    
    
    
     
    <button
      class="extension-link svelteui-c-jlBcFc svelteui-Button-root svelteui-c-jlBcFc-cWGvDy-variation-subtle svelteui-c-jlBcFc-iPJLV-css svelte-5xpm5q"
      tabindex="0"
    >
       
      BetaExtension
       
       
       
    </button>
    
    
    
     
    <button
      class="extension-link svelteui-c-jlBcFc svelteui-Button-root svelteui-c-jlBcFc-cWGvDy-variation-subtle svelteui-c-jlBcFc-iPJLV-css svelte-5xpm5q"
      tabindex="0"
    >
       
      ZebraExtension
       
       
       
    </button>
    
  </nav>
</div>
`;

exports[`Extensions > should render only extension errors (no visible links) 1`] = `
<div>
  <nav
    aria-label="Extensions"
    class="svelte-4iy04k"
  />
</div>
`;

exports[`Extensions > should render single extension info 1`] = `
<div>
  <nav
    aria-label="Extensions"
    class="svelte-4iy04k"
  >
    
    
     
    <button
      class="extension-link svelteui-c-jlBcFc svelteui-Button-root svelteui-c-jlBcFc-cWGvDy-variation-subtle svelteui-c-jlBcFc-iPJLV-css svelte-5xpm5q"
      tabindex="0"
    >
       
      TestExtension
       
       
       
    </button>
    
  </nav>
</div>
`;
